File: Java/DiffieHellman/DiffieHellman.java
public class DiffieHellman {

    public static void main(String[] args) {
        int p = 11; // Prime number
        int g = 2; // Generator of p
        int a = 9; // Alice's private value
        int b = 4; // Bob's private value

        int pubA = (int) Math.pow(g, a) % p; // Alice's public key
        int pubB = (int) Math.pow(g, b) % p; // Bob's public key

        int priA = (int) Math.pow(pubB, a) % p; // Final key that Alice got
        int priB = (int) Math.pow(pubA, b) % p; // Final key that Bob got

        System.out.println("Alice's key :: " + priA + "and Bob's key :: " + priB);
    }
}

File: algorithms/Go/DiffieHellman/DiffieHellman.go
package main

import (
	"math"
	"math/rand"
)

//https://de.wikipedia.org/wiki/Diffie-Hellman-Schl%C3%BCsselaustausch#/media/Datei:Diffie-Hellman-Schl%C3%BCsselaustausch.svg
func diffiehellman() (int, int) {
	p := 9999971 //Large prime number
	g := 4       //Natural number, smaller than p

	//Alice secret
	a := rand.Intn(p)
	//Bobs secret
	b := rand.Intn(p)

	//Alice public key
	A := int(math.Pow(float64(g), float64(a))) % p
	//Bobs public key
	B := int(math.Pow(float64(g), float64(b))) % p

	/*
		Transmit A to Bob
		Transmit B to Alice
	*/

	//Shared key, calculated by Alice
	Ka := int(math.Pow(float64(B), float64(a))) % p
	//Shared key, calculated by Bob
	Kb := int(math.Pow(float64(A), float64(b))) % p

	return Ka, Kb
}
